### DISCUSSION TITLE
Dynamic linking not supported in `create` command - deploy the library contract first, then provide the address to link at compile time

### DISCUSSION THREAD

**Discussion by evronm** [foundry-rs/foundry#1618] [Help] [answered]
This is the error message I got from forge create.  Here's the weird thing:  My tests run fine, and I could swear I've used forge create with a library before.  So, questions:

1.  Is there anything I can do to my code to make it work as is?

2.  If not, Where do I provide the address to the compiled library?  Also, Can I use forge to compile it or do I need to use e.g. solc?

Thanks in advance.

**Comment by onbjerg** [comment]
The error message tells you what to do: Deploy the library contract and then link it at compile time :) `forge test` does this linking dynamically for you.

You need to specify `--libraries` to `forge create` see the reference for more info: https://book.getfoundry.sh/reference/forge/forge-create.html

**Comment by evronm** [comment]
Thanks for the quick reply!  

That said, this really defeats the purpose.  I went from a 2 line script to compile, deploy and initialize my contracts to...  Well, I'm not exactly sure (dealing with solc?) but I guess I'll figure it out.  Are there any plans to support dynamic linking in the future?

**Reply by onbjerg** [reply]
I'm not sure I understand - this would only add one extra `forge create` and an additional flag, which shouldn't be 1 line? You don't need to deal with solc directly 🤔 

Dynamic linking will likely come in the future, but really, if you use libraries the best path is to specify the library address yourself, otherwise that defeats the purpose of using a library in the first place 😄  The main concept of libraries is that people would deploy 1 canonical version of the library and use that, reducing the amount of code on-chain. Reality is a bit different - I don't think people really do this, but that's the idea anyway

**Comment by evronm** [comment]
Well, problem is, I get the same message when trying to deploy the library, so I assumed I had to do that manually.  Here's the code:  https://github.com/evronm/Attributions/blob/e030c5fa13f18246c11d4fff6983177bdce6126b/src/Common.sol and here's the command I used:

forge create --legacy --rpc-url $RPC_URL --private-key $KEY ./src/Common.sol:Kvs

Do I need to specify all the libraries in the file?  If so, how?  Or do I need to split the libraries into separate files?

I really appreciate all the help.

**Reply by onbjerg** [reply]
Yes, if your libraries use libraries you need to specify those too. You can specify them in `foundry.toml` (refer to the book). For example, your Kvs library uses your Utils library, so you need to deploy that first

**Comment by evronm** [comment]
OK, I'm SO close now!  I think my remaining issue is how to specify multiple libraries on the command line (I really don't want to mess with config files).  I've tried the following:

--libraries $UTILS_LIB $KVMS_LIB $STR_ADDRS_LIB $TAGS_LIB
--libraries $UTILS_LIB --libraries $KVMS_LIB --libraries $STR_ADDRS_LIB --libraries $TAGS_LIB
--libraries "[$UTILS_LIB $KVMS_LIB $STR_ADDRS_LIB $TAGS_LIB]"

The last one gives:  "failed to parse, too many arguments passed: Utils" which I thought was interesting.  What am I doing wrong?  When I figure this out, I'm pretty sure I'm all set.

**Comment by evronm** [comment, marked as answer]
NVM, figured it out. FWIW, one of my variables was set incorrectly.  The correct syntax is a --libraries for each library.

Thanks again for all your help!

**Comment by evronm** [comment]
Crap!  I hate to reopen this, but I've made a few changes in my code, none of which involve any new libraries, and I'm getting the same error again.  I've checked and all the libraries are accounted for in my script.  Is there any way of figuring out what libraries the compiler is expecting?  The docs don't say anything about a verbose option or logging, but is there a different way?  Thanks again.

**Reply by onbjerg** [reply]
I think the latest Forge should tell you what libraries are missing

**Reply by evronm** [reply]
Not AFAICT.  I updated using npm, and it's still not showing me which libraries it expects.  Thankfully remix compiles the project just fine, but I'd really like to go back to my foundry-only workflow (much more streamlined).

**Reply by onbjerg** [reply]
Updated through npm? If you updated `@foundry-rs/hardhat` that does not necessarily update Foundry, just that package. To update to the latest Foundry you must use `foundryup`

**Reply by evronm** [reply]
Yup, that did it.  Thanks yet again.
